src/tests/linalg/test_linalg.f90

```diff
@@ -0,0 +1,436 @@
+    write(*,*) "test_diag_int32"
+    mask = reshape([((merge(.true.,.false.,i==j+1), i=1,n), j=1,n)], [n,n])
+    a = 0
+    a = unpack([1_int32,1_int32],mask,a)
+    call check(all(diag([1,1],-1) == a), &
+          msg="all(diag([1,1],-1) == a) failed.", warn=warn)
+    call check(all(diag([1,1],1) == transpose(a)), &
+          msg="all(diag([1,1],1) == transpose(a)) failed.", warn=warn)
+  end subroutine
+  subroutine test_diag_int64
+    integer, parameter :: n = 4
+    integer(int64) :: a(n,n), c(0:2*n-1)
+    logical :: mask(n,n)
+    integer :: i, j
+    
+    write(*,*) "test_diag_int64"
+
+    mask = reshape([((merge(.true.,.false.,i+1==j), i=1,n), j=1,n)], [n,n])
+    a = 0
+    a = unpack([1_int64,1_int64,1_int64],mask,a)
+
+    call check(all(diag([1,1,1],1) == a), &
+          msg="all(diag([1,1,1],1) == a) failed.", warn=warn)
+    call check(all(diag([1,1,1],-1) == transpose(a)), &
+          msg="all(diag([1,1,1],-1) == transpose(a)) failed.", warn=warn)
+
+
+    ! Fill array c with Catalan numbers
+    do i = 0, 2*n-1
+      c(i) = catalan_number(i)
+    end do
+    ! Symmetric Hankel matrix filled with Catalan numbers (det(H) = 1)
+    do i = 1, n
+      do j = 1, n
+        a(i,j) = c(i-1 + (j-1))
+      end do
+    end do
+    call check(all(diag(a,-2) == diag(a,2)), &
+          msg="all(diag(a,-2) == diag(a,2))", warn=warn)
+  end subroutine
+
+
+
+
+  subroutine test_trace_rsp
+    integer, parameter :: n = 5
+    real(sp) :: a(n,n)
+    integer :: i
+    write(*,*) "test_trace_rsp"
+    a = reshape([(i,i=1,n**2)],[n,n])
+    call check(abs(trace(a) - sum(diag(a))) < epsilon(1.0_sp), &
+          msg="abs(trace(a) - sum(diag(a))) < epsilon(1.0_sp) failed.",warn=warn)
+  end subroutine
+
+  subroutine test_trace_rsp_nonsquare
+    integer, parameter :: n = 4
+    real(sp) :: a(n,n+1), ans
+    integer :: i
+    write(*,*) "test_trace_rsp_nonsquare"
+
+    ! 1   5   9  13  17
+    ! 2   6  10  14  18
+    ! 3   7  11  15  19
+    ! 4   8  12  16  20
+    a = reshape([(i,i=1,n*(n+1))],[n,n+1])
+    ans = sum([1._sp,6._sp,11._sp,16._sp])
+
+    call check(abs(trace(a) - ans) < epsilon(1.0_sp), &
+          msg="abs(trace(a) - ans) < epsilon(1.0_sp) failed.",warn=warn)
+  end subroutine
+
+  subroutine test_trace_rdp
+    integer, parameter :: n = 4
+    real(dp) :: a(n,n)
+    integer :: i
+    write(*,*) "test_trace_rdp"
+    a = reshape([(i,i=1,n**2)],[n,n])
+    call check(abs(trace(a) - sum(diag(a))) < epsilon(1.0_dp), &
+          msg="abs(trace(a) - sum(diag(a))) < epsilon(1.0_dp) failed.",warn=warn)
+  end subroutine
+
+  subroutine test_trace_rdp_nonsquare
+    integer, parameter :: n = 4
+    real(dp) :: a(n,n-1), ans
+    integer :: i
+    write(*,*) "test_trace_rdp_nonsquare"
+
+    !  1   25   81
+    !  4   36  100  
+    !  9   49  121  
+    ! 16   64  144  
+    a = reshape([(i**2,i=1,n*(n-1))],[n,n-1])
+    ans = sum([1._dp,36._dp,121._dp])
+
+    call check(abs(trace(a) - ans) < epsilon(1.0_dp), &
```

Comment:
When submitting tests for `mean` and `var`, it was mentioned that `epsilon()` is too strict. Therefore, you should multiply it by e.g., 1000 (see [here](https://github.com/fortran-lang/stdlib/blob/c95f7a9a6d031f9871e09564432ebcdd7035c179/src/tests/stats/test_mean.f90#L9) for the current implementation) to avoid any issues.